discard string context of git config's core.excludesFile

Radvendii opened the pull request:
Before this change, any store path in your git config would become part of the string context of any of the parsed strings, and then prevent us from appending it to a path.

Upon further reflection, we don't care even if `core.excludesFile` is a store path itself, so we can unconditionally throw away the string context.

I think this closes #65

Radvendii committed:
discard string context of git config's core.excludesFile

Before this change, any store path in your git config would become part
of the string context of any of the parsed strings, and then prevent us
from appending it to a path.

Upon further reflection, we don't care even if core.excludesFile is a
store path itself, so we can unconditionally throw away the string
context.

roberth commented on find-files.nix:
```diff
@@ -242,7 +242,10 @@ rec {
           (guard (toLower section == "core" && toLower key == "excludesfile"))
           (_:
-            resolveFile (home /.) value
+            # SAFETY: Ultimately this path will get passed into
+            # `builtins.readFile`, which ignores the context anyways. The
+            # context is only relevant if it becomes part of a derivation.
```
`readFile` does realise the context, and resolves CA outputs (`ca-derivations`), so this isn't quite true.

Maybe something like this, though it could be more detailed.

```suggestion
            # SAFETY: In some circumstances, `value` carries a string context,
            # which would result in an error in `resolveFile (home /.)`.
```

Radvendii replied:
Ah! I had tested out that the output of `readFile` doesn't  change depending on the context of its input, but I didn't think about the mechanics of how it gets the result.

Let me know if the new comment is okay.

roberth reviewed (changes requested):
This project can only exist by virtue of its test suite, so we're going to need a regression test for a fix like this.

Radvendii committed:
fix comment

It turns out readFile does realise the context, so it could make a
difference. I would still argue that's not our responsibility and your
git config should be pointing ot paths that are already valid.

Radvendii commented:
> This project can only exist by virtue of its test suite, so we're going to need a regression test for a fix like this.

I'm trying to add a test, but it requires me to set `HOME` or `XDG_CONFIG_HOME` in order to point the impure part of the code to a config the test controls. But I can't find where I would set this up. Where does `nix/ci.nix` get used?

roberth commented:
It's a non-flake set of derivations that gets evaluated by Hercules CI, basically like `nix-build nix/ci.nix`.

You could use `listingViaNixGitignore` as a template for evaluating inside the sandbox and then use that to provide a new test case in the big attrset of `tests/default.nix`.

Radvendii committed:
WIP: add regression test for #71

Radvendii commented:
Arghhh okay so I wrote a test. The issue is that `builtins.readFile` is apparently finicky and fails to grab the context when run inside the nix sandbox like the tests are set up.

I pushed what I have, but it doesn't properly catch the bug (on master).

roberth commented:
I think the problem with the test is that it uses the wrong store. In the sandbox, `/nix/store` is just a directory and not a proper store.
I briefly gave it a try a couple days ago to get it to work with a different `storeDir` (not just relocated, but properly different; no `/nix/store` in sight), but didn't get it to work yet because the settings system for this is still kind of a mess.
I have no hope of improving that upstream until there's more progress on the grand topic of [settings](https://github.com/NixOS/nix/issues/5638), but I know it is possible to massage Nix the right way and accept its new `storeDir`...

roberth committed:
Fix guardNonEmptyString

One of the simplest mistakes.
I use more TDD nowadays...

roberth committed:
Make regression-config-with-store-path work

The tricky bit was that we had to use a different storeDir because
the normal store is not a real store in the sandbox.
Also HOME apprently wasn't picked up, and XDG_CONFIG_HOME was broken;
see previous commit.
